### PR TITLE
SAMZA-1226: relax type parameters in MessageStream functions

### PR DESCRIPTION
**Pull request by nickpan47**
relax the type parameter in user supplied functions in fluent API

### CODE REVIEW

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java [hunk 1]
```diff
@@ -82,17 +82,19 @@
    * non-stream systems (e.g., an external database).
    *
    * @param sinkFn the function to send messages in this stream to an external system
+   * @return this {@link MessageStream} object
```
Comment (jmakes) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: This is not as detailed as I would need if I were the user. 

Is "this" stream before or after the external stream? Are there any concerns that should be taken into account when using it for chaining? What if I want to consume the external stream, say, for modularity/testing purposes?

Same comment for the return param of sink() below.

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java [hunk 2]
```diff
@@ -47,20 +47,20 @@
...
    * Applies the provided 1:n function to transform a message in this {@link MessageStream}
    * to n messages in the transformed {@link MessageStream}
    *
    * @param flatMapFn the function to transform a message to zero or more messages
-   * @param <TM> the type of messages in the transformed {@link MessageStream}
+   * @param <OM> the type of messages in the output {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <TM> MessageStream<TM> flatMap(FlatMapFunction<M, TM> flatMapFn);
+  <OM> MessageStream<OM> flatMap(FlatMapFunction<? super M, ? extends OM> flatMapFn);
```
Comment (prateekm) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: Issue: Why do these functions need `? super M` as the input type? Seems like this is implying that we can provide any super-class of M (which is effectively Object) as the input. That doesn't seem right. Should this be `? extends M`?

Also, shouldn't the return type also be `? extends OM` for the operators themselves?

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java (hunk 1 above)
Comment (prateekm) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: Documentation: "object" is redundant.

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java [hunk 3]
```diff
@@ -82,17 +82,19 @@
    * non-stream systems (e.g., an external database).
    *
    * @param sinkFn the function to send messages in this stream to an external system
+   * @return this {@link MessageStream} object
    */
-  void sink(SinkFunction<M> sinkFn);
+  MessageStream<M> sink(SinkFunction<? super M> sinkFn);
```
Comment (prateekm) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: Issue: I think we decided to make these terminal because sink and sendTo can repartition the output stream while this returns the un-repartitioned stream.

File: samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java [hunk 4]
```diff
@@ -119,11 +121,21 @@ private Windows() { }
...
    * @return the created {@link Window} function.
    */
-  public static <M, K, WV> Window<M, K, WV> keyedTumblingWindow(Function<M, K> keyFn, Duration interval,
-                                                                Supplier<WV> initialValue, FoldLeftFunction<M, WV> foldFn) {
+  public static <M, K, WV> Window<M, K, WV> keyedTumblingWindow(Function<? super M, ? extends K> keyFn, Duration interval,
+                                                                Supplier<? extends WV> initialValue, FoldLeftFunction<? super M, WV> foldFn) {
 
     Trigger<M> defaultTrigger = new TimeTrigger<>(interval);
-    return new WindowInternal<M, K, WV>(defaultTrigger, initialValue, foldFn, keyFn, null, WindowType.TUMBLING);
+    return new WindowInternal<>(defaultTrigger, initialValue::get, new FoldLeftFunction<M, WV>() {
```
Comment (prateekm) [inline samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java]: Issue: Why create wrapper functions here?

File: samza-api/src/main/java/org/apache/samza/task/TaskContext.java [hunk 5]
```diff
@@ -58,10 +58,9 @@
   /**
    * Method to allow user to return customized context
```
Comment (prateekm) [inline samza-api/src/main/java/org/apache/samza/task/TaskContext.java]: Documentation: Shouldn't refer to user in third person in public APIs. Also, "Method to" is redundant.

File: samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java [hunk 6]
```diff
@@ -61,16 +61,22 @@ public StreamGraphImpl(ApplicationRunner runner, Config config) {
...
-        streamSpec -> new InputStreamInternalImpl<>(this, streamSpec, msgBuilder));
+        streamSpec -> new InputStreamInternalImpl<K, V, M>(this, streamSpec, msgBuilder));
   }
 
   @Override
   public <K, V, M> OutputStream<K, V, M> getOutputStream(String streamId,
-      Function<M, K> keyExtractor, Function<M, V> msgExtractor) {
+      Function<? super M, ? extends K> keyExtractor, Function<? super M, ? extends V> msgExtractor) {
+    if (keyExtractor == null || msgExtractor == null) {
+      throw new IllegalArgumentException("Neither keyExtractor nor msgExtractor can be null for an output stream.");
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java]: Thanks for adding these. 
Issue: Should specify which is null.

File: samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpecs.java [hunk 7]
```diff
@@ -115,8 +117,20 @@ public void init(Config config, TaskContext context) {
    * @return  the {@link StreamOperatorSpec}
    */
   public static <M, OM> StreamOperatorSpec<M, OM> createStreamOperatorSpec(
-      FlatMapFunction<M, OM> transformFn, MessageStreamImpl<OM> nextStream, int opId) {
-    return new StreamOperatorSpec<>(transformFn, nextStream, OperatorSpec.OpCode.FLAT_MAP, opId);
+      FlatMapFunction<? super M, ? extends OM> transformFn, MessageStreamImpl<OM> nextStream, int opId) {
+    return new StreamOperatorSpec<>(new FlatMapFunction<M, OM>() {
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpecs.java]: Issue: Why create wrapper functions?

File: samza-core/src/main/java/org/apache/samza/operators/stream/IntermediateStreamInternalImpl.java [hunk 8]
```diff
@@ -33,13 +33,13 @@
...
-  public IntermediateStreamInternalImpl(StreamGraphImpl graph, StreamSpec streamSpec,
-      Function<M, K> keyExtractor, Function<M, V> msgExtractor, BiFunction<K, V, M> msgBuilder) {
+  public IntermediateStreamInternalImpl(StreamGraphImpl graph, StreamSpec streamSpec, Function<? super M, ? extends K> keyExtractor,
+      Function<? super M, ? extends V> msgExtractor, BiFunction<? super K, ? super V, ? extends M> msgBuilder) {
     super(graph);
     this.streamSpec = streamSpec;
-    this.keyExtractor = keyExtractor;
-    this.msgExtractor = msgExtractor;
-    this.msgBuilder = msgBuilder;
+    this.keyExtractor = keyExtractor::apply;
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/stream/IntermediateStreamInternalImpl.java]: Issue: Seems unusual, this is already a functional interface . Why is this required?

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java (hunk 1 above)
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: This means before the external stream. For consuming from the external stream, it should either be partitionBy or sendThrough.

File: samza-api/src/main/java/org/apache/samza/application/StreamApplication.java [hunk 9]
```diff
@@ -24,11 +24,46 @@
 
 
 /**
- * This interface defines a template for stream application that user will implement to create operator DAG in {@link StreamGraph}.
+ * This interface defines a template for stream application that user will implement to initialize operator DAG in {@link StreamGraph}.
```
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/application/StreamApplication.java]: I always thought that the javadoc needs to refer to the readers as the third person. No strong opinion on this though. Only one thing that if we want to make this call, let's be consistent across the code base.

File: samza-api/src/main/java/org/apache/samza/application/StreamApplication.java [hunk 10]
```diff
@@ -24,11 +24,46 @@
...
+ *   public static void main(String[] args) {
+ *     CommandLine cmdLine = new CommandLine();
+ *     Config config = cmdLine.loadConfig(cmdLine.parser().parse(args));
+ *     PageViewCounterExample userApp = new PageViewCounterExample();
+ *     userApp.initBlackList(config);
+ *     ApplicationRunner localRunner = ApplicationRunner.getLocalRunner(config);
+ *     localRunner.run(userApp);
+ *   }
+ *
+ *   private void initBlackList(Config config) {
```
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/application/StreamApplication.java]: Let me change the example s.t. it does not include the init phase.

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java [hunk 11]
```diff
@@ -47,20 +47,20 @@
    * transformed {@link MessageStream}.
    *
    * @param mapFn the function to transform a message to another message
-   * @param <TM> the type of messages in the transformed {@link MessageStream}
+   * @param <OM> the type of messages in the output {@link MessageStream}
```
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: Sure. I will leave the type parameter name as TM as well then.

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java (hunk 2 above)
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: Refer to out discussion earlier on PR #51 , the whole purpose of this PR is to relax the function type parameter, not the MessageStream type parameter. The purpose is to relax the restriction for user to provide their own functions to be applied to the MessageStream&lt;M&gt;. Hence, here are the basic rules I applied in this change:
1. No type change to the types we define (i.e. MessageStream, StreamGraph, OperatorSpec, Window, etc)
2. Relaxed the type parameters of functions user supplies s.t. the user supplied functions can be applied to the message types in MessageStream.

That's why {code}? super M{code} and {code}? extends M{code} are defined as the type parameters for the functions, not to the MessageStream class. Restricting the type to M totally defeat the whole purpose of this PR.

File: samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java (hunk 4 above)
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java]: This is to formulate the type of Function that our internal class will depend on. Leaking the captured type parameter to internal functions leads to a whole lot changes over the full implementation stack, which does not seem to be necessary for the purpose (i.e. just allow relaxed type parameter from user-supplied functions).

File: samza-api/src/main/java/org/apache/samza/task/TaskContext.java (hunk 5 above)
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/task/TaskContext.java]: +1 if you want to make a full scale blanket change on this.

File: samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java (hunk 6 above)
Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java]: Will add to the error message.

File: samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpecs.java (hunk 7 above)
Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpecs.java]: Same answer as earlier.

File: samza-core/src/main/java/org/apache/samza/operators/stream/IntermediateStreamInternalImpl.java (hunk 8 above)
Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/operators/stream/IntermediateStreamInternalImpl.java]: Same as the reason for wrapper function.

File: samza-api/src/main/java/org/apache/samza/task/TaskContext.java (hunk 5 above)
Comment (prateekm) [inline samza-api/src/main/java/org/apache/samza/task/TaskContext.java]: Will do. Only mentioned this here since TaskContext isn't part of the operators package so will be easy to miss.

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java (hunk 3 above)
Comment (prateekm) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: I think it'll be safer and more clearly explainable to leave them as terminal. Users can still chain explicitly on the previous message stream if they wish to. Don't feel strongly about it, up to you.

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java (hunk 2 above)
Comment (prateekm) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: @nickpan47 explained offline. The way to read this is that the user can provide a function that takes a super-type of M as input since it's always safe to cast the actual provided M to its super-type.

E.g., if M == Integer, this signature allows users to provide a more general FilterFunction<Number> instead of a FilterFunction<Integer> if that's what they want to use.

File: samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java (hunk 4 above)
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java]: Talked offline with @prateekm  and we agreed that it should be safe to simply do a type-cast here. Will make the change in several places with the same wrapper.

File: samza-api/src/main/java/org/apache/samza/operators/MessageStream.java (hunk 3 above)
Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/operators/MessageStream.java]: Sold on this one. Will keep sink / sendTo terminal. The implementation in PR is more like a side output. Maybe we can add later as sideSink()/sideSend(), if more and more use cases popping up.

File: samza-core/src/main/java/org/apache/samza/operators/stream/IntermediateStreamInternalImpl.java [hunk 12]
```diff
@@ -33,13 +33,13 @@
   private final Function<M, V> msgExtractor;
   private final BiFunction<K, V, M> msgBuilder;
 
-  public IntermediateStreamInternalImpl(StreamGraphImpl graph, StreamSpec streamSpec,
-      Function<M, K> keyExtractor, Function<M, V> msgExtractor, BiFunction<K, V, M> msgBuilder) {
+  public IntermediateStreamInternalImpl(StreamGraphImpl graph, StreamSpec streamSpec, Function<? super M, ? extends K> keyExtractor,
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/stream/IntermediateStreamInternalImpl.java]: A little cleaner to do these casts in StreamGraph and leave these types as K, V, M. Both here and InputStreamInternalImpl.

File: samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpecs.java [hunk 13]
```diff
@@ -53,7 +53,7 @@ private OperatorSpecs() {}
    * @return  the {@link StreamOperatorSpec}
    */
   public static <M, OM> StreamOperatorSpec<M, OM> createMapOperatorSpec(
-      MapFunction<M, OM> mapFn, MessageStreamImpl<OM> nextStream, int opId) {
+      MapFunction<? super M, ? extends OM> mapFn, MessageStreamImpl<OM> nextStream, int opId) {
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpecs.java]: Will be a little cleaner to do these casts in MessageStreamImpl and leave these as K, V, M

File: samza-api/src/main/java/org/apache/samza/application/StreamApplication.java [hunk 14]
```diff
@@ -24,11 +24,44 @@
...
+ *       .filter(m -> !(m.getMessage().getEventTime() < System.currentTimeMillis() - MAX_TIMEOUT))
+ *       .sendTo(pageViewEventFilteredStream);
+ *   }
+ *
+ *   // local execution mode
+ *   public static void main(String[] args) {
+ *     CommandLine cmdLine = new CommandLine();
+ *     Config config = cmdLine.loadConfig(cmdLine.parser().parse(args));
+ *     PageViewCounterExample userApp = new PageViewCounterExample();
+ *     userApp.initBlackList(config);
```
Comment (navina) [inline samza-api/src/main/java/org/apache/samza/application/StreamApplication.java]: What does userApp.initBlackList(config) mean?

Comment (nickpan47) [inline samza-api/src/main/java/org/apache/samza/application/StreamApplication.java]: Urh... Good catch!!! That should be gone. I am removing it and committing.

Thanks!
